### PR TITLE
Kinetic Surface Reconstruction

### PR DESCRIPTION
**Pull request by soesau**
Kinetic Surface Reconstruction

Complementing the Kinetic Shape Partition package with the reconstruction pipeline described in the original paper.

* Affected package(s): Kinetic Shape Partitioning and Kinetic Surface Reconstruction
* Issue(s) solved (if any):
* Feature/Small Feature (if any):
* Link to compiled documentation: [doc](https://cgal.github.io/7908/v0/Manual/index.html)
* License and copyright ownership:  GeometryFactory/Inria

### CODE REVIEW

File: Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/basic.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+using Normal_map   = typename Point_set::Vector_map;
+
+using KSR = CGAL::Kinetic_surface_reconstruction_3<Kernel, Point_set, Point_map, Normal_map>;
+
+
+int main(const int argc, const char** argv) {
+  // Input.
+  std::string filename = CGAL::data_file_path("points_3/building.ply");
+
+  Point_set point_set;
```
Comment (afabri) [inline Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/basic.cpp]: Use a more high level `CGAL::read` function.

File: Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/basic.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  auto param = CGAL::parameters::maximum_distance(0.5)
+    .maximum_angle(10)
+    .k_neighbors(12)
+    .minimum_region_size(50);
+
+  // Algorithm.
+  KSR ksr(point_set, param);
+
+  ksr.detect_planar_shapes(param);
```
Comment (afabri) [inline Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/basic.cpp]: No need to call all the steps individually.

File: Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/building.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
+#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
+#include <CGAL/Kinetic_surface_reconstruction_3.h>
+#include <CGAL/Point_set_3.h>
+#include <CGAL/Point_set_3/IO.h>
+#include <CGAL/IO/PLY.h>
+#include <CGAL/IO/polygon_soup_io.h>
+
+#include "include/Parameters.h"
```
Comment (afabri) [inline Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/building.cpp]: Are these used?

Comment (afabri) [reply Kinetic_surface_reconstruction/examples/Kinetic_surface_reconstruction/building.cpp]: Now I get it.   The parameters must be added properly.

File: Kinetic_surface_reconstruction/doc/Kinetic_surface_reconstruction/Kinetic_surface_reconstruction.txt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+- \ref Chapter_Kinetic_Shape_Partition "Kinetic shape partition": *k*, *reorient_bbox* and *bbox_dilation_ratio*
+
+The reconstruction adds two new parameters:
+- *external_nodes*:
+The graph cut formulation embeds a vertex into each volume and connects all vertices if the corresponding volumes share a common face. While each face inside the kinetic shape partition is exactly between two volumes, faces on the boundary have to be considered. Thus, 6 external vertices are inserted into the graph representing a volume on each side of the bounding box. The parameter *external_nodes* allows the user to either provide a fixed label *inside* or *outside* for each node or leave the label up to the energy minimization. Typical choices for this parameter are to choose all external nodes as outside for scanned objects, e.g., as in \cgalFigureRef{Kspr_pipelinefig}, or the *ZMIN* node *inside* and all other nodes *outside* as for exterior scans of buildings. The default value for this parameter is to leave the labels of the external nodes up to the energy minimization.
+
+- \f$\lambda\f$:
```
Comment (afabri) [inline Kinetic_surface_reconstruction/doc/Kinetic_surface_reconstruction/Kinetic_surface_reconstruction.txt]: Here it should not be the greek letter lambda.

File: Kinetic_surface_reconstruction/doc/Kinetic_surface_reconstruction/Kinetic_surface_reconstruction.txt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+The \f$\lambda\f$ parameter trades the data faithfulness of the energy minimization for low complexity. The parameter has to be choosen in the range of \f$[0,1)\f$. 0 indicates the only the voting of input points are considered in the energy minimization, while the default value of \f$0,5\f$ gives an equal weight to data faithfulness and low complexity. The value should be choosen according to the quality of the input data. If the point cloud is accurate, low in noise and complete a low value can be chosen.
+
+\section ksrParameterChoice Choice of parameters
```
Comment (afabri) [inline Kinetic_surface_reconstruction/doc/Kinetic_surface_reconstruction/Kinetic_surface_reconstruction.txt]:
```suggestion
\section ksrParameterChoice Choice of Parameters
```

Comment (afabri) [reply Kinetic_surface_reconstruction/doc/Kinetic_surface_reconstruction/Kinetic_surface_reconstruction.txt]: check that please for all headings

File: Kinetic_shape_partition/include/CGAL/KSP_3/Finalizer.h [hunk 6]
```diff
@@ -116,11 +116,6 @@ class Finalizer {
 
     create_volumes();
```
Comment (afabri) [inline Kinetic_shape_partition/include/CGAL/KSP_3/Finalizer.h]: The files which are not part of the public API should go in a directory and namespace `KSP_3/internal`

File: Kinetic_shape_partition/include/CGAL/KSP_3/Graphcut.h [hunk 7]
```diff
@@ -52,7 +52,7 @@ namespace KSR_3 {
     using Delaunay_3 = CGAL::Delaunay_triangulation_3<Kernel>;
```
Comment (afabri) [inline Kinetic_shape_partition/include/CGAL/KSP_3/Graphcut.h]: What is the role of the Delaunay  triangulation in the graphcut?

Comment (afabri) [reply Kinetic_shape_partition/include/CGAL/KSP_3/Graphcut.h]: In the Graphcut file I also see `double min = 10000000000;`   which should probably be replaced with `std::numeric_limits<double>`

File: Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h [hunk 8]
```diff
@@ -10,8 +10,8 @@
 //
 // Author(s)     : Sven Oesau, Florent Lafarge, Dmitry Anisimov, Simon Giraudot
 
-#ifndef CGAL_KINETIC_SHAPE_RECONSTRUCTION_3_H
-#define CGAL_KINETIC_SHAPE_RECONSTRUCTION_3_H
+#ifndef CGAL_KINETIC_SURFACE_RECONSTRUCTION_3_H
+#define CGAL_KINETIC_SURFACE_RECONSTRUCTION_3_H
 
 #include <CGAL/license/Kinetic_shape_partition.h>
```
Comment (afabri) [inline Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h]: Generate license file for the package.

File: Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h [hunk 9]
```diff
@@ -40,57 +40,61 @@
 
 namespace CGAL
 {
-#ifndef DOXYGEN_RUNNING
 /*!
-* \ingroup PkgKineticShapePartition
-  \brief Piece-wise linear reconstruction via inside/outside labeling of a kinetic partition using graph cut.
+* \ingroup PkgKineticSurfaceReconstructionRef
+  \brief Reconstruction pipeline for piece-wise planar surface reconstruction from a point cloud via inside/outside labeling of a kinetic partition using graph cut.
```
Comment (afabri) [inline Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h]:
```suggestion
  \brief Reconstruction pipeline for piecewise planar surface reconstruction from a point cloud via inside/outside labeling of a kinetic partition using graph cut.
```

File: Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h [hunk 10]
```diff
@@ -40,57 +40,61 @@
...
 
   \tparam GeomTraits
-    must be a model of `KineticPartitionTraits`.
+    must be a model of `KineticShapePartitionTraits_3`.
+
+  \tparam PointSet
```
Comment (afabri) [inline Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h]: Should become `PointRange`.   Making this comment looks like a déjà-vu.

File: Kinetic_shape_partition/include/CGAL/KSP_3/Graphcut.h (hunk 7 above)
Comment (soesau) [reply Kinetic_shape_partition/include/CGAL/KSP_3/Graphcut.h]: Delaunay was a left over and has been removed and the second suggestion was integrated.

File: Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h [hunk 11]
```diff
@@ -233,12 +271,224 @@ class Kinetic_shape_reconstruction_3 {
   @return
   created kinetic partition data structure
 
-  \pre `successful partition`
+  \pre `partition created`
```
Comment (afabri) [inline Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h]: Why is created backticked??

File: Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h [hunk 12]
```diff
@@ -233,12 +271,224 @@ class Kinetic_shape_reconstruction_3 {
...
+  The z axis is considered as vertical upwards pointing.
+
+  \tparam OutputPointIterator
+  an output iterator taking Point_3.
+
+  \tparam OutputPolygonIterator
+  an output iterator taking polygon indices std::vector<std::size_t>.
+
+  \param beta
+  trades the impact of the data term for impact of the regularization term. Should be in the range [0, 1).
```
Comment (afabri) [inline Kinetic_surface_reconstruction/include/CGAL/Kinetic_surface_reconstruction_3.h]:
```suggestion
  trades the impact of the data term for impact of the regularization term. Must be in the range `[0, 1)`.
```

**Pull request closed without merging**